File: java/RemoveDuplicatesFromSortedArrayII.java
/*
 * Filename: RemoveDuplicatesFromSortedArrayII.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Mon 02 Mar 2015 07:58:19 PM EST
 */

public class Solution {
	public int removeDuplicates(int[] A) {
		int length = A.length;
		if (length < 3) return length;
		else {
			int curr = 2;
			int prev = 1;
			int res = 2;

			for (; curr < length; curr++) {
				if (A[curr] != A[prev] || A[curr] != A[prev-1]) {
					res++;
					A[++prev] = A[curr];
				}
			}

			return res;
		}
	}
}

File: cpp/remove_duplicates_from_sorted_array_ii.cc
/*
 * Filename: remove_duplicates_from_sorted_array_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 02 Aug 2015 08:44:44 PM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		int removeDuplicates(vector<int>& nums)
		{
			if (nums.size() < 3)
			{
				return nums.size();
			}

			int i = 2;
			int res = 1;

			while (i < nums.size())
			{
				if (nums[res] - nums[i] || nums[res - 1] - nums[i])
				{
					nums[++res] = nums[i];
				}

				i++;
			}

			return res + 1;
		}
};
